samples/matter/common/src/watchdog/watchdog.h

```diff
@@ -0,0 +1,187 @@
+	 *    argument. If the callback is provided it will be called in each feed interval time periodically.
+	 */
+	class WatchdogSource : public sys_snode_t {
+	public:
+		/**
+		 * @brief Callback for feeding purposes
+		 *
+		 * This callback will be run when feedingInterval timeout occurs.
+		 * It must call the Feed method within the specific context.
+		 * If this callback cannot be call within the CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT timeout the
+		 * reboot occurs.
+		 *
+		 * For example to feed an Watchdog source dedicated for Main thread
+		 * you can use Nrf::PostTask method:
+		 *
+		 *  Nrf::PostTask([watchdog] { watchdog->Feed(); });
+		 */
+		using FeedingCallback = void (*)(WatchdogSource *watchdogSource);
+
+		/**
+		 * @brief Create new Watchdog feeding source
+		 *
+		 * Each Watchdog source may contain a callback which is called when the feedingInterval time is elapsed.
+		 * If at least one of the watchdog source is not fed within the
+		 * CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT then the reboot occurs.
+		 *
+		 * @param FeedingInterval if the callback argument is provided then after elapsing time in milliseconds
+		 * of this value the callback will be set.
+		 * @param callback FeedingCallback to be called in the specific place to reset the Watchdog timer
+		 */
+		WatchdogSource(uint32_t feedingInterval = 0, FeedingCallback callback = nullptr)
+			: mFeedingTime(feedingInterval), mCallback(callback)
+		{
+		}
+
+		/**
+		 * @brief Remove the Watchdog source
+		 *
+		 * The destructor perform a several steps to remove a WatchdogSource:
+		 * 1. Disables the Watchdog peripheral.
+		 * 2. Removes this Watchdog source.
+		 * 3. Restores all other Watchdog sources that are already registered.
+		 * 4. Enables the Watchdog peripheral.
+		 */
+		~WatchdogSource();
+
+		/**
+		 * @brief Reset the Watchdog timer from a specific source.
+		 *
+		 * If the Watchdog peripheral doesn't receive feed signal within the
+		 * CONFIG_NCS_SAMPLE_MATTER_WATCHDOG_TIMEOUT then the device reboot occurs.
+		 *
+		 * @note This method can be called within various context, wherever you want to detect high time usage,
+		 * or infinite loops. For example you can call this method within a specific thread.
+		 */
+		void Feed();
+
+		friend bool Enable();
+		friend bool Disable();
+		friend bool InstallSource(WatchdogSource &source);
+
+	private:
+		constexpr static uint8_t kInvalidChannel = -1;
+
+		/* Pause Watchdog when the CPU is in sleep state or is halted by the debugger */
+		constexpr static uint8_t kWatchdogOptions = WDT_OPT_PAUSE_HALTED_BY_DBG | WDT_OPT_PAUSE_IN_SLEEP;
+
+		/**
+		 * @brief Restore the Watchdog source
+		 *
+		 * This method must be called in order to re-enable a Watchdog source if the Enable() static method has
+		 * been called.
+		 *
+		 * @warning This method must be called before calling Enable() if the Disable function has been called.
```

Comment:
I'd expect that if this method is private, then it is called in by the class internally as a part of implementation (not API).

Reply:
I see that it can be public, since the user would like to disable only several sources, but to do this they need to disable all of them and then call Install only on the specific objects.